Description: All previous patches
  This patch contains all of the previous patches up to Debian revision version 1.0b2-11.  Needed in debian/patches after upgrading debian/compat from 4 to 9. 


--- wmtime-1.0b2.orig/wmgeneral/wmgeneral.c
+++ wmtime-1.0b2/wmgeneral/wmgeneral.c
@@ -12,6 +12,8 @@
 	---
 	CHANGES:
 	---
+	10/10/2003 (Simon Law, sfllaw@debian.org)
+		* changed the parse_rcfile function to use getline instead of fgets.
 	02/05/1998 (Martijn Pieterse, pieterse@xs4all.nl)
 		* changed the read_rc_file to parse_rcfile, as suggester by Marcelo E. Magallon
 		* debugged the parse_rc file.
@@ -21,6 +23,7 @@
 
 */
 
+#define _GNU_SOURCE
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
@@ -43,6 +46,11 @@ int			screen;
 int			x_fd;
 int			d_depth;
 XSizeHints	mysizehints;
+/* Deal with strange X11 function prototyping...
+ * If I don't do this, I will get warnings about the sign
+ * of the width/height variables - which are thrown away anyway. */
+int dummy_int_width, dummy_int_height;
+unsigned int uint_width, uint_height;
 XWMHints	mywmhints;
 Pixel		back_pix, fore_pix;
 char		*Geometry = "";
@@ -83,17 +91,18 @@ int CheckMouseRegion(int, int);
 void parse_rcfile(const char *filename, rckeys *keys) {
 
 	char	*p;
-	char	temp[128];
+	char	*line = NULL;
+	size_t  line_size = 0;
 	char	*tokens = " :\t\n";
 	FILE	*fp;
 	int		i,key;
 
 	fp = fopen(filename, "r");
 	if (fp) {
-		while (fgets(temp, 128, fp)) {
+		while (getline(&line, &line_size, fp) >= 0) {
 			key = 0;
 			while (key >= 0 && keys[key].label) {
-				if ((p = strstr(temp, keys[key].label))) {
+				if ((p = strstr(line, keys[key].label))) {
 					p += strlen(keys[key].label);
 					p += strspn(p, tokens);
 					if ((i = strcspn(p, "#\n"))) p[i] = 0;
@@ -270,6 +279,7 @@ void openXwindow(int argc, char *argv[],
 	unsigned int	borderwidth = 1;
 	XClassHint		classHint;
 	char			*display_name = NULL;
+	char			*geometry = NULL;
 	char			*wname = argv[0];
 	XTextProperty	name;
 
@@ -282,7 +292,9 @@ void openXwindow(int argc, char *argv[],
 
 	for (i=1; argv[i]; i++) {
 		if (!strcmp(argv[i], "-display")) 
-			display_name = argv[i+1];
+			display_name = argv[++i];
+		else if (!strcmp(argv[i], "-geometry"))
+			geometry = argv[++i];
 	}
 
 	if (!(display = XOpenDisplay(display_name))) {
@@ -307,16 +319,21 @@ void openXwindow(int argc, char *argv[],
 	fore_pix = GetColor("black");
 
 	XWMGeometry(display, screen, Geometry, NULL, borderwidth, &mysizehints,
-				&mysizehints.x, &mysizehints.y,&mysizehints.width,&mysizehints.height, &dummy);
-
-	mysizehints.width = 64;
-	mysizehints.height = 64;
+	            &mysizehints.x, &mysizehints.y,
+	            &dummy_int_width, &dummy_int_height, &dummy);
+	if (geometry)
+		XParseGeometry(geometry, &mysizehints.x, &mysizehints.y,
+		               &uint_width, &uint_height);
+
+   /* Override width/height anyway */
+	uint_width = 64;
+	uint_height = 64;
 		
 	win = XCreateSimpleWindow(display, Root, mysizehints.x, mysizehints.y,
-				mysizehints.width, mysizehints.height, borderwidth, fore_pix, back_pix);
+				uint_width, uint_height, borderwidth, fore_pix, back_pix);
 	
 	iconwin = XCreateSimpleWindow(display, win, mysizehints.x, mysizehints.y,
-				mysizehints.width, mysizehints.height, borderwidth, fore_pix, back_pix);
+				uint_width, uint_height, borderwidth, fore_pix, back_pix);
 
 	/* Activate hints */
 	XSetWMNormalHints(display, win, &mysizehints);
@@ -364,3 +381,6 @@ void openXwindow(int argc, char *argv[],
 	XMapWindow(display, win);
 
 }
+
+/* vim: ts=4 columns=82
+ */
--- wmtime-1.0b2.orig/wmtime/Makefile
+++ wmtime-1.0b2/wmtime/Makefile
@@ -4,12 +4,16 @@ OBJS =	wmtime.o \
 		../wmgeneral/wmgeneral.o \
 		../wmgeneral/misc.o \
 		../wmgeneral/list.o
+XPMS = wmtime-master.xpm wmtime-mask.xbm
+
+CC = gcc
+CFLAGS = -O2
 
 .c.o:
-	cc -c -O2 -Wall $< -o $*.o
+	$(CC) $(CFLAGS) -c -Wall $< -o $*.o
 
-wmtime: $(OBJS)
-	cc -o wmtime $^ $(LIBDIR) $(LIBS)
+wmtime: $(OBJS) $(XPMS)
+	$(CC) $(CFLAGS) -o wmtime $(OBJS) $(LIBDIR) $(LIBS)
 
 clean::
 	for i in $(OBJS) ; do \
--- /dev/null
+++ wmtime-1.0b2/wmtime/wmtime.1
@@ -0,0 +1,178 @@
+'\" t
+.\" Man page for wmtime
+.\" Copyright (c) 2003  Software in the Public Interest, Inc.
+.\"
+.\" This program is free software; you can redistribute it and/or modify
+.\" it under the terms of the GNU General Public License as published by
+.\" the Free Software Foundation; either version 2 of the License, or (at
+.\" your option) any later version.
+.\"
+.\" This program is distributed in the hope that it will be useful, but
+.\" WITHOUT ANY WARRANTY; without even the implied warranty of
+.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+.\" General Public License for more details.
+.\"
+.\" You should have received a copy of the GNU General Public License
+.\" along with this program; if not, write to the Free Software
+.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
+.\"
+.TH WMTIME 1 "May 1998" "WMTIME 1.0b2"
+
+.SH NAME
+wmtime \- Window Maker dockapp that displays the time and date
+
+.SH SYNOPSIS
+
+.B wmtime
+[
+.I OPTION
+]
+
+.SH DESCRIPTION
+
+.B WMTime
+displays the time and date and gives you some nice additional
+features too. It is intended for docking in Window Maker.
+
+It currently provides:
+
+.IP \(bu 4
+the time and date;
+
+.IP \(bu
+a realtime morphing interface (analog <> digital mode);
+.IP \(bu
+auto\-scaled and anti\-aliased hands;
+.IP \(bu
+localization, displaying the day and date in various languages;
+.IP \(bu
+three user\-defined commands to launch.
+
+.SH OPTIONS
+
+.TP
+.BR \-d , \ \-digital
+Display the digital clock on startup, instead of the analog clock.
+
+.TP
+.BI \-display \ display
+This option specifies the X server to contact; see
+.IR X(7x) .
+
+.TP
+.BI \-geometry \ geometry
+This option specifies the preferred position of clock; see
+.IR X(7x) .
+
+.TP
+.BR \-n , \ \-noseconds
+Disables the second hand.
+
+.TP
+.B \-h
+Show help information.
+
+.TP
+.B \-v
+Print the version number.
+
+.SH USAGE
+The
+.B WMTime
+window is separated into top and bottom sections.  The top section
+contains a clock and the bottom section contains the date.
+
+The clock can be toggled between analog and digital modes.  To do this,
+click in the bottom section of the window.
+
+WMTime can also be used to launch programs.  You may click either left,
+middle, or right mouse buttons in the top section of the window.  The
+pre\-configured program will be launched according to the mouse button
+clicked.  (see
+.B CONFIGURATION FILE
+below.)
+
+In order to move
+.BR WMTime ,
+drag on the outer edges.  These are not sensitive to mouse clicks.
+
+.SH "DOCKING IN WINDOW MANAGERS"
+
+.TP
+Window Maker
+
+Window Maker users should drag and drop the
+.B WMTime
+window on the Dock.  Then, right\-click on the border of the window and
+select \*(lqSettings...\*(rq.  Check \*(lqStart when Window Maker
+is started\*(rq.
+
+.TP
+AfterStep
+
+AfterStep users should put the following in their
+.I $HOME/.steprc
+
+.RS 16
+Wharf wmtime \- MaxSwallow "wmtime" wmtime &
+.RE
+
+.TP
+Other window managers
+
+.B WMTime
+runs nicely as a 64x64 shaped icon on your desktop.
+
+.SH "CONFIGURATION FILE"
+
+.B WMTime
+can launch three user\-defined commands, triggered by left, middle and
+right mouse button clicks.  You can define the commands to launch in
+.I $HOME/.wmtimerc
+
+.RS
+.PD 0
+left: xterm
+.PP
+middle: xload
+.PP
+right: xcalc
+.PP
+.PD 
+.RE
+
+If
+.B WMTime
+detects a
+.I $HOME/.wmtimerc
+file, it will launch the appropriate command when you click on the clock.
+
+The system administrator can define default commands in
+.IR /etc/wmtimerc .
+The administrator may also choose to \*(lqfix\*(rq particular commands,
+making it impossible for users to change.  These commands can be defined in
+.IR /etc/wmtimerc.fixed ,
+although this isn't a nice thing to do.
+
+.SH FILES
+
+.I /etc/wmtimerc
+.br
+.I $HOME/.wmtimerc
+.br
+.I /etc/wmtimerc.fixed
+
+.SH AUTHORS
+
+.B WMTime
+was written by Martijn Pieterse and Antoine Nulle.
+
+This manual page was written by Simon Law <sfllaw@debian.org> for the
+.B Debian
+system (but may be used by others). It is based on the documentation provided
+by the original program.
+
+This manual is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 2 of the License, or (at
+your option) any later version.
--- wmtime-1.0b2.orig/wmtime/wmtime-master.xpm
+++ wmtime-1.0b2/wmtime/wmtime-master.xpm
@@ -29,44 +29,44 @@ static char * wmtime_master_xpm[] = {
 "    X......................................................o        X......................................................o    ................................",
 "    X.........................OOO..........................o        X......................................................o    ................................",
 "    X.........................OOO+.........................o        X......................................................o    ................................",
-"    X..............@@.........OOO+........@@...............o        X...++++...++++.......++++...++++.......++++...++++....o    ................................",
-"    X..............@@+.........+++........@@+..............o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
-"    X...............++.....................++..............o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
-"    X......................................................o        X..+....+.+....+.....+....+.+....+.....+....+.+....+...o    ................................",
+"    X..............@@.........OOO+........@@...............o        X......................................................o    ................................",
+"    X..............@@+.........+++........@@+..............o        X......................................................o    ................................",
+"    X...............++.....................++..............o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    X......@@......................................@@......o        X......................................................o    ................................",
+"    X......@@+.....................................@@+.....o        X......................................................o    ................................",
+"    X.......++......................................++.....o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
 "    X......................................................o        X...++++...++++.......++++...++++.......++++...++++....o    ................................",
-"    X......................................................o        X..+....+.+....+.....+....+.+....+.....+....+.+....+...o    ................................",
-"    X......@@......................................@@......o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
-"    X......@@+.....................................@@+.....o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
-"    X.......++......................................++.....o        X...++++...++++.......++++...++++.......++++...++++....o    ................................",
+"    X......................................................o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
+"    X......................................................o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
+"    X..OOO...........................................OOO...o        X..+....+.+....+.....+....+.+....+.....+....+.+....+...o    ................................",
+"    X..OOO+..........................................OOO+..o        X...++++...++++.......++++...++++.......++++...++++....o    ................................",
+"    X..OOO+..........................................OOO+..o        X..+....+.+....+.....+....+.+....+.....+....+.+....+...o    ................................",
+"    X...+++...........................................+++..o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
+"    X......................................................o        X..+....+.+....+..+..+....+.+....+..+..+....+.+....+...o    ................................",
+"    X......................................................o        X...++++...++++.......++++...++++.......++++...++++....o    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
 "    X......................................................o        X......................................................o    ................................",
 "    X......................................................o        X......................................................o    ................................",
+"    X......@@......................................@@......o        X......................................................o    ................................",
+"    X......@@+.....................................@@+.....o        X......................................................o    ................................",
+"    X.......++......................................++.....o        X......................................................o    ................................",
 "    X......................................................o        X......................................................o    ................................",
 "    X......................................................o        X......................................................o    ................................",
-"    X......................................................o        oooooooooooooooooooooooooooooooooooooooooooooooooooooooo    ................................",
-"    X......................................................o                                                                    ................................",
-"    X..OOO...........................................OOO...o                                                                    ................................",
-"    X..OOO+..........................................OOO+..o                                                                    ................................",
-"    X..OOO+..........................................OOO+..o                                                                    ................................",
-"    X...+++...........................................+++..o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......@@......................................@@......o                                                                    ................................",
-"    X......@@+.....................................@@+.....o                                                                    ................................",
-"    X.......++......................................++.....o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    X..............@@.....................@@...............o                                                                    ................................",
-"    X..............@@+........OOO.........@@+..............o                                                                    ................................",
-"    X...............++........OOO+.........++..............o                                                                    ................................",
-"    X.........................OOO+.........................o                                                                    ................................",
-"    X..........................+++.........................o                                                                    ................................",
-"    X......................................................o                                                                    ................................",
-"    oooooooooooooooooooooooooooooooooooooooooooooooooooooooo                                                                    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    X..............@@.....................@@...............o        X......................................................o    ................................",
+"    X..............@@+........OOO.........@@+..............o        X......................................................o    ................................",
+"    X...............++........OOO+.........++..............o        X......................................................o    ................................",
+"    X.........................OOO+.........................o        X......................................................o    ................................",
+"    X..........................+++.........................o        X......................................................o    ................................",
+"    X......................................................o        X......................................................o    ................................",
+"    oooooooooooooooooooooooooooooooooooooooooooooooooooooooo        oooooooooooooooooooooooooooooooooooooooooooooooooooooooo    ................................",
 "                                                                                                                                ................................",
 "                                                                                                                                ................................",
 "    XXXXXXXXXXXXXXXX  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX        XXXXXXXXXXXXXXXX  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX    ................................",
--- wmtime-1.0b2.orig/wmtime/wmtime.c
+++ wmtime-1.0b2/wmtime/wmtime.c
@@ -26,6 +26,13 @@
 	----
 	Changes:
 	----
+   15/07/2008 (Paul Harris, harris.pc@gmail.com)
+      * Minor changes to correct build warnings
+	09/10/2003 (Simon Law, sfllaw@debian.org)
+		* Add -geometry support
+		* Add -noseconds support
+		* Make the digital clock fill the space provided
+		* Eliminated exploitable static buffers
 	17/05/1998 (Antoine Nulle, warp@xs4all.nl)
 		* Updated version number and some other minor stuff
 	16/05/1998 (Antoine Nulle, warp@xs4all.nl)
@@ -50,6 +57,7 @@
 		* Copied the source from wmmon.
 */
 
+#define _GNU_SOURCE
 #include <stdlib.h>
 #include <stdio.h>
 #include <time.h>
@@ -76,9 +84,9 @@
  /* Defines */
 /***********/
 
-#define LEFT_ACTION (NULL)
-#define MIDDLE_ACTION (NULL)
-#define RIGHT_ACTION (NULL)
+const char* default_left_action = NULL;
+const char* default_middle_action = NULL;
+const char* default_right_action = NULL;
 
 #define WMMON_VERSION "1.0b2"
 
@@ -86,55 +94,63 @@
  /* Global Variables */
 /********************/
 
-char	*ProgName;
 int		digital = 0;
+int		noseconds = 0;
 char	day_of_week[7][3] = { "SU", "MO", "TU", "WE", "TH", "FR", "SA" };
-char	mon_of_year[12][4] = { "JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOW", "DEC" };
+char	mon_of_year[12][4] = { "JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC" };
 
 /* functions */
-void usage(void);
+void usage(char *);
 void printversion(void);
 
 void wmtime_routine(int, char **);
 void get_lang();
 
-void main(int argc, char *argv[]) {
+int main(int argc, char *argv[]) {
 
 	int		i;
-	
-
-	/* Parse Command Line */
+	char    *name = argv[0];
 
-	ProgName = argv[0];
-	if (strlen(ProgName) >= 6)
-		ProgName += (strlen(ProgName) - 6);
-	
 	for (i=1; i<argc; i++) {
 		char *arg = argv[i];
 
 		if (*arg=='-') {
 			switch (arg[1]) {
 			case 'd' :
-				if (strcmp(arg+1, "display") && strcmp(arg+1, "digital")) {
-					usage();
-					exit(1);
+				if (strcmp(arg+1, "display")
+				    && strcmp(arg+1, "digital") && strcmp(arg+1, "d")) {
+					usage(name);
+					return 1;
 				}
-				if (!strcmp(arg+1, "digital"))
+				if (!strcmp(arg+1, "digital") || !(strcmp(arg+1, "d")))
 					digital = 1;
 				break;
+			case 'g' :
+				if (strcmp(arg+1, "geometry")) {
+					usage(name);
+					return 1;
+				}
+				break;
+			case 'n' :
+				if (strcmp(arg+1, "noseconds") && strcmp(arg+1, "n")) {
+					usage(name);
+					return 1;
+				} else {
+					noseconds = 1;
+				}
+				break;
 			case 'v' :
 				printversion();
-				exit(0);
-				break;
+				return 0;
 			default:
-				usage();
-				exit(0);
-				break;
+				usage(name);
+				return 1;
 			}
 		}
 	}
 	get_lang();
 	wmtime_routine(argc, argv);
+	return 0;
 }
 
 /************/
@@ -143,19 +159,22 @@ void main(int argc, char *argv[]) {
 void get_lang(){
    FILE *fp;
    int i;
-   char temp[5];
+   const int line_size = 5;
+   char line[line_size];
    
    fp=fopen("language","r");
    if (fp) {
+	   /* Grab the days of the week */
        for (i=0;i<7;i++){
-	   fgets(temp,4,fp);
-	   strncpy(day_of_week[i],temp,2);
+		   fgets(line, line_size, fp);
+		   strncpy(day_of_week[i], line, 2);
        };
+	   /* Grab the names of the months */
        for (i=0;i<12;i++){
-	   fgets(temp,5,fp);
-	   strncpy(mon_of_year[i],temp,3);
+		   fgets(line, line_size, fp);
+		   strncpy(mon_of_year[i], line, 3);
        };
-   	fclose(fp);
+	   fclose(fp);
     };
 }
 
@@ -191,39 +210,38 @@ void wmtime_routine(int argc, char **arg
 	long		curtime;
 	long		nexttime;
 
-	char		temp[128];
-	char		*p;
+	char		*conffile = NULL;
 
 	/* Scan through ~/.wmtimerc for the mouse button actions. */
-	if (LEFT_ACTION) left_action = strdup(LEFT_ACTION);
-	if (MIDDLE_ACTION) middle_action = strdup(MIDDLE_ACTION);
-	if (RIGHT_ACTION) right_action = strdup(RIGHT_ACTION);
-
-	/* Scan throught  the .rc files */
-	strcpy(temp, "/etc/wmtimerc");
-	parse_rcfile(temp, wmtime_keys);
-
-	p = getenv("HOME");
-	strcpy(temp, p);
-	strcat(temp, "/.wmtimerc");
-	parse_rcfile(temp, wmtime_keys);
+	if (default_left_action) left_action = strdup(default_left_action);
+	if (default_middle_action) middle_action = strdup(default_middle_action);
+	if (default_right_action) right_action = strdup(default_right_action);
 
-	strcpy(temp, "/etc/wmtimerc.fixed");
-	parse_rcfile(temp, wmtime_keys);
+	/* Scan through the .rc files */
+	if (asprintf(&conffile, "/etc/wmtimerc") >= 0) {
+		parse_rcfile(conffile, wmtime_keys);
+		free(conffile);
+	}
 
+	if (asprintf(&conffile, "%s/.wmtimerc", getenv("HOME")) >= 0) {
+		parse_rcfile(conffile, wmtime_keys);
+		free(conffile);
+	}
+
+	if (asprintf(&conffile, "/etc/wmtimerc.fixed") >= 0) {
+		parse_rcfile(conffile, wmtime_keys);
+		free(conffile);
+	}
 
 	openXwindow(argc, argv, wmtime_master_xpm, wmtime_mask_bits, 128, 64);
 
-	copyXPMArea(0, 0, 128, 64, 0, 98);
-	if (digital) {
-		copyXPMArea(64, 0, 64, 64, 0, 0);
-		setMaskXY(-64, 0);
-	} else {
-		copyXPMArea(0, 0, 64, 64, 64, 0);
-		setMaskXY(0, 0);
-	}
+	/* Mask out the right parts of the clock */
+	copyXPMArea(0, 0, 128, 64, 0, 98);   /* Draw the borders */
+	copyXPMArea(0, 0, 64, 64, 64, 0);    /* Draw the clock face */
+	copyXPMArea(64, 98, 64, 64, 0, 0);   /* Draw the LCD background */
+	setMaskXY(0, 0);
 
-	/* add mouse region */
+    /* add mouse region */
 	AddMouseRegion(0, 5, 48, 58, 60);
 	AddMouseRegion(1, 5, 5, 58, 46);
 
@@ -282,12 +300,10 @@ void wmtime_routine(int argc, char **arg
 
 						if (digital) {
 							copyXPMArea(64, 98, 64, 64, 0, 0);
-							setMaskXY(-64, 0);
 							DrawTime(time_struct->tm_hour, time_struct->tm_min, time_struct->tm_sec);
 							DrawDate(time_struct->tm_wday, time_struct->tm_mday, time_struct->tm_mon);
 						} else {
-							copyXPMArea(0, 98, 64, 64, 0, 0);
-							setMaskXY(0, 0);
+							copyXPMArea(0, 98, 64, 64, 0, 0);	
 							DrawWijzer(time_struct->tm_hour, time_struct->tm_min, time_struct->tm_sec);
 							DrawDate(time_struct->tm_wday, time_struct->tm_mday, time_struct->tm_mon);
 						}
@@ -324,24 +340,28 @@ void wmtime_routine(int argc, char **arg
 \*******************************************************************************/
 
 void DrawTime(int hr, int min, int sec) {
-
-	
-	char	temp[16];
-	char	*p = temp;
+	const int time_size = 16;
+	char	time[time_size];
+	char	*p = time;
 	int		i,j,k=6;
 
 	/* 7x13 */
 
-	sprintf(temp, "%02d:%02d:%02d ", hr, min, sec);
+	if (noseconds) {
+		snprintf(time, time_size, "%02d:%02d ", hr, min);
+	}
+	else {
+		snprintf(time, time_size, "%02d:%02d:%02d ", hr, min, sec);
+	}
 
-	for (i=0; i<3; i++) {
+	for (i=0; i < 3; i++) {
 		for (j=0; j<2; j++) {
-			copyXPMArea((*p-'0')*7 + 1, 84, 8, 13, k, 6);
+			copyXPMArea((*p-'0')*7 + 1, 84, 8, 13, k, 18);
 			k += 7;
 			p++;
 		}
 		if (*p == ':') {
-			copyXPMArea(71, 84, 5, 13, k, 6);
+			copyXPMArea(71, 84, 5, 13, k, 18);
 			k += 4;
 			p++;
 		}
@@ -353,15 +373,15 @@ void DrawTime(int hr, int min, int sec)
 \*******************************************************************************/
 
 void DrawDate(int wkday, int dom, int month) {
-
-	
-	char	temp[16];
-	char	*p = temp;
+	const int date_size = 16;
+	char	date[date_size];
+	char	*p = date;
 	int		i,k;
 
 	/* 7x13 */
 
-	sprintf(temp, "%.2s%02d%.3s  ", day_of_week[wkday], dom, mon_of_year[month]);
+	snprintf(date, date_size,
+			"%.2s%02d%.3s  ", day_of_week[wkday], dom, mon_of_year[month]);
 
 	k = 5;
 	for (i=0; i<2; i++) {
@@ -541,6 +561,9 @@ void DrawWijzer(int hr, int min, int sec
 		}
 	}
 	/**********************************************************************/
+	if (noseconds)
+		return;    /* Skip drawing the seconds. */
+
 	psi = sec * (M_PI / 30.0);
 
 	dx = floor(sin(psi) * 22 * 0.9 + 0.5);
@@ -612,15 +635,16 @@ void DrawWijzer(int hr, int min, int sec
 |* usage																	   *|
 \*******************************************************************************/
 
-void usage(void) {
-
-	fprintf(stderr, "\nwmtime - programming: tijno, (de)bugging & design: warp, web hosting: bobby\n\n");
-	fprintf(stderr, "usage:\n");
-	fprintf(stderr, "\t-digital\tdigital clock\n");
-	fprintf(stderr, "\t-display <display name>\n");
-	fprintf(stderr, "\t-h\tthis screen\n");
-	fprintf(stderr, "\t-v\tprint the version number\n");
-	fprintf(stderr, "\n");
+void usage(char *name) {
+	printf("Usage: %s [OPTION]...\n", name);
+	printf("WindowMaker dockapp that displays the time and date.\n");
+	printf("\n");
+	printf("  -d, -digital         display the digital clock\n");
+	printf("  -display DISPLAY     contact the DISPLAY X server\n");
+	printf("  -geometry GEOMETRY   position the clock at GEOMETRY\n");
+	printf("  -n, -noseconds       disables the second hand\n");
+	printf("  -h                   display this help and exit\n");
+	printf("  -v                   output version information and exit\n");
 }
 
 /*******************************************************************************\
@@ -628,8 +652,8 @@ void usage(void) {
 \*******************************************************************************/
 
 void printversion(void) {
-
-	if (!strcmp(ProgName, "wmtime")) {
-		fprintf(stderr, "%s\n", WMMON_VERSION);
-	}
+	printf("WMTime version %s\n", WMMON_VERSION);
 }
+
+/* vim: sw=4 ts=4 columns=82
+ */
